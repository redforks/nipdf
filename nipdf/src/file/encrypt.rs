use crate::object::ObjectId;
use ahash::HashMap;
use arc4::Arc4;
use log::error;
use md5::{Digest, Md5};
use nipdf_macro::{pdf_object, TryFromIntObject, TryFromNameObject};
use prescript::{sname, Name};
use tinyvec::{Array, ArrayVec, TinyVec};

#[derive(TryFromIntObject, Default, Debug, PartialEq, Eq, Clone, Copy)]
pub enum Algorithm {
    #[default]
    Undocument = 0,
    Key40 = 1,
    Key40AndMore = 2,
    Unpublished = 3,
    DefinedInDoc = 4,
}

#[derive(TryFromIntObject, PartialEq, Eq, Debug, Clone, Copy, PartialOrd, Ord)]
pub enum StandardHandlerRevion {
    V2 = 2,
    V3 = 3,
    V4 = 4,
}

#[derive(TryFromNameObject, PartialEq, Eq, Debug, Clone, Copy, PartialOrd, Ord, Default)]
pub enum DecryptMethod {
    #[default]
    None,
    V2,
    AesV2,
}

const fn identity() -> Name {
    sname("Identity")
}

#[pdf_object(Some("CryptFilter"))]
pub trait CryptFilterDictTrait {
    #[key("CFM")]
    #[try_from]
    fn decrypt_method(&self) -> DecryptMethod;
}

#[pdf_object(())]
pub trait EncryptDictTrait {
    fn filter(&self) -> Name;

    fn sub_filter(&self) -> Option<Name>;

    #[or_default]
    #[key("V")]
    #[try_from]
    fn algorithm(&self) -> Algorithm;

    #[key("Length")]
    #[default(40)]
    fn key_length(&self) -> u32;

    #[key("P")]
    fn permission_flags(&self) -> u32;

    #[key("R")]
    #[try_from]
    fn revison(&self) -> StandardHandlerRevion;

    /// 32-byte long string.
    #[key("O")]
    fn owner_password_hash(&self) -> &[u8];

    /// 32-byte long string.
    #[key("U")]
    fn user_password_hash(&self) -> &[u8];

    #[key("CF")]
    #[one_or_more]
    #[nested]
    fn crypt_filter_params(&self) -> HashMap<Name, CryptFilterDict>;

    /// Crypt filter used if Crypt field not set on Stream dictionary
    #[key("StmF")]
    #[default_fn(identity)]
    fn stream_default_crypt_filter(&self) -> Name;

    /// Crypt filter used to decode strings
    #[key("StrF")]
    #[default_fn(identity)]
    fn string_crypt_filter(&self) -> Name;
}

impl<'a, 'b> EncryptDict<'a, 'b> {
    /// Return the default crypt filter for stream and string.
    /// If `self.>revision()` not V4, return (Identity, Identity).
    /// Stream default crypt filter: lookup `self.crypt_filter_params` by
    /// `self.stream_default_crypt_filter()`, If not found, use Identity, otherwise by
    /// `decrypt_method()` return value.
    /// String default crypt filter similar to stream, but lookup by
    /// `self.string_default_crypt_filter()`. If any error occurs, uses Identity.
    pub fn default_filters(&self) -> (CryptFilter, CryptFilter) {
        use anyhow::{anyhow, Result};
        fn resolve_filter(
            name: &Name,
            params: &HashMap<Name, CryptFilterDict>,
        ) -> Result<CryptFilter> {
            if name == &identity() {
                return Ok(CryptFilter::Identity);
            }

            params
                .get(name)
                .ok_or_else(|| anyhow!("crypt filter params not found: {}", name))
                .and_then(|d| d.decrypt_method())
                .map(|m| match m {
                    DecryptMethod::None => CryptFilter::Identity,
                    DecryptMethod::V2 => CryptFilter::Rc4,
                    DecryptMethod::AesV2 => CryptFilter::Aes,
                })
        }

        fn _do(this: &EncryptDict) -> Result<(CryptFilter, CryptFilter)> {
            if this.revison()? != StandardHandlerRevion::V4 {
                return Ok((CryptFilter::Identity, CryptFilter::Identity));
            }

            let params = this.crypt_filter_params()?;
            let stream_filter = resolve_filter(&this.stream_default_crypt_filter()?, &params)?;
            let string_filter = resolve_filter(&this.string_crypt_filter()?, &params)?;
            Ok((stream_filter, string_filter))
        }

        if !matches!(
            self.algorithm().unwrap(),
            Algorithm::Key40 | Algorithm::Key40AndMore,
        ) {
            todo!("Algorithm: {:?}", self.algorithm().unwrap());
        }

        _do(self).unwrap_or_else(|e| {
            error!("failed to get default crypt filter, use Identity: {}", e);
            (CryptFilter::Identity, CryptFilter::Identity)
        })
    }
}

const PADDING: [u8; 32] = [
    0x28u8, 0xBF, 0x4E, 0x5E, 0x4E, 0x75, 0x8A, 0x41, 0x64, 0x00, 0x4E, 0x56, 0xFF, 0xFA, 0x01,
    0x08, 0x2E, 0x2E, 0x00, 0xB6, 0xD0, 0x68, 0x3E, 0x80, 0x2F, 0x0C, 0xA9, 0xFE, 0x64, 0x53, 0x69,
    0x7A,
];

/// Pad or truncate the password to 32 bytes.
/// If the password is longer than 32 bytes, the extra bytes are ignored.
/// If the password is shorter than 32 bytes, it is padded with bytes
/// [28 BF 4E 5E 4E 75 8A 41 64 00 4E 56 FF FA 01 08 2E 2E 00 B6 D0 68 3E 80 2F 0C A9 FE 64 53 69
/// 7A]
fn pad_trunc_password(s: &[u8]) -> [u8; 32] {
    let mut iter = s.iter().copied().chain(PADDING).take(32);
    std::array::from_fn(|_| iter.next().unwrap())
}

// algorithm 2
pub fn calc_encrypt_key(
    revion: StandardHandlerRevion,
    key_length: usize,
    user_password: &[u8],
    owner_hash: &[u8; 32],
    permission_flags: u32,
    doc_id: &[u8],
) -> Box<[u8]> {
    let user_password = pad_trunc_password(user_password);
    let mut md5 = Md5::new();
    md5.update(user_password);
    md5.update(owner_hash);
    md5.update(permission_flags.to_le_bytes());
    md5.update(doc_id);
    // md5.update(&[0xff, 0xff, 0xff, 0xff]);
    let mut hash = md5.finalize();
    let n = key_length / 8;
    if revion > StandardHandlerRevion::V2 {
        for _ in 0..50 {
            hash = Md5::digest(&hash[..n]);
        }
    }
    (&hash[..n]).into()
}

// algorithm 4 and 5
fn calc_user_hash(
    revion: StandardHandlerRevion,
    key_length: usize,
    user_password: &[u8],
    owner_hash: &[u8; 32],
    permission_flags: u32,
    doc_id: &[u8],
) -> [u8; 32] {
    let key = calc_encrypt_key(
        revion,
        key_length,
        user_password,
        owner_hash,
        permission_flags,
        doc_id,
    );

    if revion == StandardHandlerRevion::V2 {
        let mut r = PADDING.to_owned();
        Arc4::with_key(&key).encrypt(&mut r);
        r
    } else {
        let mut md5 = Md5::new();
        md5.update(PADDING);
        md5.update(doc_id);
        let mut hash = md5.finalize();
        let mut tmp = key.to_vec();
        for i in 0..=19 {
            for (t, k) in tmp.as_mut_slice().iter_mut().zip(&key[..]) {
                *t = *k ^ i;
            }
            Arc4::with_key(&tmp[..]).encrypt(&mut hash);
        }
        let mut r = [0u8; 32];
        r[..16].copy_from_slice(&hash[..]);
        r
    }
}

// algorithm 6
pub fn authorize_user(
    revion: StandardHandlerRevion,
    key_length: usize,
    user_password: &[u8],
    owner_hash: &[u8; 32],
    user_hash: &[u8; 32],
    permission_flags: u32,
    doc_id: &[u8],
) -> bool {
    let hash = calc_user_hash(
        revion,
        key_length,
        user_password,
        owner_hash,
        permission_flags,
        doc_id,
    );

    if revion == StandardHandlerRevion::V2 {
        &hash[..] == user_hash
    } else {
        hash[..16] == user_hash[..16]
    }
}

/// algorithm 3 step a to d
#[allow(dead_code)]
fn calc_rc4_key(
    revion: StandardHandlerRevion,
    key_length: usize,
    owner_password: &[u8],
) -> Box<[u8]> {
    let mut owner_password = Md5::digest(pad_trunc_password(owner_password));
    if revion > StandardHandlerRevion::V2 {
        for _ in 0..50 {
            owner_password = Md5::digest(&owner_password[..]);
        }
    }
    (&owner_password[..(key_length / 8)]).into()
}

// algorithm 7
#[allow(dead_code)]
fn authorize_owner(
    revion: StandardHandlerRevion,
    key_length: usize,
    owner_password: &[u8],
    owner_hash: &[u8; 32],
    user_hash: &[u8; 32],
    permission_flag: u32,
    doc_id: &[u8],
) -> bool {
    let rc4_key = &calc_rc4_key(revion, key_length, owner_password);
    let mut decrypt = owner_hash.to_vec();
    if revion == StandardHandlerRevion::V2 {
        Arc4::with_key(rc4_key).encrypt(&mut decrypt);
    } else {
        let mut tmp = rc4_key.to_vec();
        for i in (0..19).rev() {
            for (t, k) in tmp.as_mut_slice().iter_mut().zip(&rc4_key[..]) {
                *t = *k ^ i;
            }
            Arc4::with_key(&tmp[..]).encrypt(&mut decrypt);
        }
    }

    authorize_user(
        revion,
        key_length,
        &decrypt,
        owner_hash,
        user_hash,
        permission_flag,
        doc_id,
    )
}

pub trait VecLike {
    fn drain(&mut self, range: std::ops::Range<usize>) -> ArrayVec<[u8; 16]>;
    fn as_mut_slice(&mut self) -> &mut [u8];
}

pub trait Decryptor {
    fn new(key: &[u8], id: ObjectId) -> Self;
    fn decrypt<V: VecLike>(&self, data: &mut V);
}

impl<A> VecLike for TinyVec<A>
where
    A: Array<Item = u8>,
{
    fn drain(&mut self, range: std::ops::Range<usize>) -> ArrayVec<[u8; 16]> {
        self.drain(range).collect()
    }

    fn as_mut_slice(&mut self) -> &mut [u8] {
        self.as_mut_slice()
    }
}

impl VecLike for Vec<u8> {
    fn drain(&mut self, range: std::ops::Range<usize>) -> ArrayVec<[u8; 16]> {
        self.drain(range).collect()
    }

    fn as_mut_slice(&mut self) -> &mut [u8] {
        self.as_mut_slice()
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CryptFilter {
    Identity,
    Rc4,
    Aes,
}

pub struct Rc4Decryptor(ArrayVec<[u8; 16]>);

impl Decryptor for Rc4Decryptor {
    fn new(key: &[u8], id: ObjectId) -> Self {
        let n = key.len();
        let mut k = TinyVec::<[u8; 16 + 5]>::with_capacity(n + 5);
        k.extend_from_slice(key);
        k.extend_from_slice(&id.id().0.to_le_bytes()[..3]);
        k.extend_from_slice(&id.generation().to_le_bytes()[..]);
        let key = Md5::digest(&k[..]);
        let key = key.into_iter().take((n + 5).min(16)).collect();
        Self(key)
    }

    fn decrypt<V: VecLike>(&self, data: &mut V) {
        Arc4::with_key(&self.0).encrypt(data.as_mut_slice());
    }
}

pub struct AesDecryptor(ArrayVec<[u8; 16]>);

impl Decryptor for AesDecryptor {
    fn new(key: &[u8], id: ObjectId) -> Self {
        let n = key.len();
        let mut k = TinyVec::<[u8; 16 + 5 + 4]>::with_capacity(n + 5 + 4);
        k.extend_from_slice(key);
        k.extend_from_slice(&id.id().0.to_le_bytes()[..3]);
        k.extend_from_slice(&id.generation().to_le_bytes()[..]);
        k.extend_from_slice(b"sAlT");
        let key = Md5::digest(&k[..]);
        let key = key.into_iter().take((n + 5).min(16)).collect();
        Self(key)
    }

    /// Decode data using aes, work in cbc mode,block size is 16 or Aes128.
    /// the initialization vector is a 16-byte random  number that is stored as the first 16 bytes
    /// of the encrypted data.
    /// Pad the data using the PKCS#5 padding scheme.
    fn decrypt<V: VecLike>(&self, data: &mut V) {
        use aes::cipher::{block_padding::Pkcs7, BlockDecryptMut, KeyIvInit};
        type Aes128CbcDec = cbc::Decryptor<aes::Aes128>;

        let iv_data = data.drain(0..16);
        let mut iv = [0u8; 16];
        iv.copy_from_slice(&iv_data[..]);
        Aes128CbcDec::new(self.0.as_ref().into(), &iv.into())
            .decrypt_padded_mut::<Pkcs7>(data.as_mut_slice())
            .unwrap();
    }
}

#[cfg(test)]
mod tests;
