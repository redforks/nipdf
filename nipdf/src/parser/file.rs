use super::{ws_terminated, FileError, ParseError, ParseResult};
use crate::{
    function::{Domain, Domains},
    object::{Dictionary, Entry, Frame, FrameSet, ObjectValueError, RuntimeObjectId, XRefSection},
    parser::{parse_dict, parse_indirect_stream},
};
use log::{error, info};
use memchr::memmem::rfind;
use nom::{
    branch::alt,
    bytes::complete::tag,
    character::complete::{char, satisfy, u32},
    combinator::{complete, map, map_res, opt, recognize},
    error::{context, ErrorKind, ParseError as NomParseError},
    multi::{count, fold_many1, many0},
    number::complete::{be_u16, be_u24, be_u32, be_u8},
    sequence::{preceded, separated_pair, tuple},
    InputIter, InputLength, InputTake, Parser, Slice,
};
use prescript::sname;
use std::{fmt::Display, ops::RangeFrom, str::from_utf8_unchecked};

/// Return `None`` if file not start with `%PDF-`
pub fn parse_header(buf: &[u8]) -> ParseResult<Option<&str>> {
    let one_digit = || satisfy(|c| c.is_ascii_digit());

    fn new_header(buf: &[u8]) -> std::result::Result<Option<&str>, FileError> {
        assert_eq!(3, buf.len());
        if buf[0] != b'1' {
            Err(FileError::UnsupportedVersion(
                String::from_utf8_lossy(buf).to_string(),
            ))
        } else {
            // safe to call from_utf8_unchecked(), because buf is checked for ascii digits
            Ok(Some(unsafe { from_utf8_unchecked(buf) }))
        }
    }

    let (buf, _) = tag("%")(buf)?;
    let (buf, v) = opt(tag("PDF-"))(buf)?;
    if v.is_none() {
        return Ok((buf, None));
    }

    map_res(
        recognize(tuple((one_digit(), char('.'), one_digit()))),
        new_header,
    )(buf)
}

/// Return start position of object tag from the end of the buffer.
/// Object tag occupies a whole line.
fn r_find_start_object_tag(mut buf: &[u8], tag: &[u8]) -> Option<usize> {
    fn is_new_line(buf: &[u8], pos: usize) -> bool {
        buf.get(pos).map_or(true, |ch| *ch == b'\n' || *ch == b'\r')
    }

    loop {
        match rfind(buf, tag) {
            None => return None,
            Some(pos) => {
                if is_new_line(buf, pos + tag.len()) && (pos == 0 || is_new_line(buf, pos - 1)) {
                    return Some(pos);
                } else {
                    buf = &buf[..pos];
                }
            }
        }
    }
}

fn new_r_to_tag<'a>(tag: &'static [u8]) -> impl FnMut(&'a [u8]) -> ParseResult<'a, ()> {
    move |buf| {
        let pos = r_find_start_object_tag(buf, tag);
        if let Some(pos) = pos {
            Ok((&buf[pos..], ()))
        } else {
            Err(nom::Err::Failure(ParseError::from_error_kind(
                buf,
                ErrorKind::Fail,
            )))
        }
    }
}

fn parse_trailer(buf: &[u8]) -> ParseResult<Dictionary> {
    preceded(ws_terminated(tag(b"trailer")), ws_terminated(parse_dict))(buf)
}

struct CrossReferenceStreamDict {
    size: u32,
    index: Option<Domains<u32>>,
    w: Vec<u32>,
}

impl CrossReferenceStreamDict {
    pub fn new(d: &Dictionary) -> Result<Self, ObjectValueError> {
        let size = d
            .get(&sname("Size"))
            .ok_or(ObjectValueError::DictKeyNotFound)?
            .int()? as u32;
        let index = d
            .get(&sname("Index"))
            .map(|o| Domains::<u32>::try_from(o).unwrap());
        let w = d
            .get(&sname("W"))
            .ok_or(ObjectValueError::DictKeyNotFound)?
            .arr()?
            .iter()
            .map(|o| o.int().map(|v| v as u32))
            .collect::<Result<Vec<_>, _>>()?;

        Ok(Self { size, index, w })
    }
}

/// Return nom parser to parse u32 value by byte length (0, 1, 2, 3, 4),
/// if n is 0, return parser takes 0 bytes and returns 0.
/// if n > 1, n32 stored in big endian n bytes.
fn segment_parser<'a, I: 'a>(n: u32) -> Box<dyn Parser<I, u32, nom::error::VerboseError<I>> + 'a>
where
    I: InputIter<Item = u8> + InputLength + InputTake + Slice<RangeFrom<usize>>,
{
    match n {
        0 => Box::new(|v| Ok((v, 0_u32))),
        1 => Box::new(map(be_u8, |v| v as u32)),
        2 => Box::new(map(be_u16, |v| v as u32)),
        3 => Box::new(be_u24),
        4 => Box::new(be_u32),
        _ => unreachable!(),
    }
}

/// Parse xref from cross-reference streams
fn parse_xref_stream(input: &[u8]) -> ParseResult<(XRefSection, Dictionary)> {
    fn to_parse_error<E: Display>(e: E) -> nom::Err<ParseError<'static>> {
        error!("should be xref table stream: {}", e);
        nom::Err::Error(ParseError::from_error_kind(b"", ErrorKind::Fail))
    }

    let (buf, stream) = parse_indirect_stream(input)?;
    let d = stream.as_dict();
    let d = CrossReferenceStreamDict::new(d).map_err(to_parse_error)?;

    let data = stream
        .decode_without_resolve_length(input, None)
        .map_err(to_parse_error)?;
    assert_eq!(3, d.w.len());
    let (a, b, c) = (d.w[0], d.w[1], d.w[2]);
    debug_assert_eq!(
        data.len() % (a + b + c) as usize,
        0,
        "stream data length should multiple of w0 + w1 + w2"
    );
    let (_, entries) = complete(count(
        move |i| tuple((segment_parser(a), segment_parser(b), segment_parser(c)))(i),
        data.len() / (a + b + c) as usize,
    ))(data.as_ref())
    .map_err(to_parse_error)?;

    let size = d.size;
    let sections = d
        .index
        .unwrap_or_else(|| Domains(vec![Domain::new(0, size)]));

    let mut r = Vec::with_capacity(sections.0.iter().map(|x| x.end as usize).sum());
    let mut entries = entries.into_iter();
    for Domain { start, end: size } in sections.0 {
        for (idx, (a, b, c)) in entries.by_ref().take(size as usize).enumerate() {
            let c: u16 = c.try_into().unwrap();
            let idx: u32 = idx.try_into().unwrap();
            match a {
                0 => r.push((start + idx, Entry::in_file(0, c, false))),
                1 => r.push((start + idx, Entry::in_file(b, c, true))),
                2 => r.push((start + idx, Entry::in_stream(RuntimeObjectId(b), c))),
                _ => info!("unknown xref stream entry type: {a}, idx: {idx}, ignored",),
            }
        }
    }

    Ok((buf, (r, stream.take_dict())))
}

// Assumes buf start from xref
fn parse_xref_table(buf: &[u8]) -> ParseResult<XRefSection> {
    let record_count_parser = context(
        "record count",
        ws_terminated(separated_pair(u32, tag(b" "), u32)),
    );
    let record_parser = context(
        "record",
        map(
            ws_terminated(tuple((
                u32,
                tag(b" "),
                u32,
                tag(b" "),
                alt((tag(b"n"), tag(b"f"))),
            ))),
            |(offset, _, generation, _, ty)| {
                // safe because clamp into u16 range
                #[allow(clippy::cast_possible_truncation)]
                Entry::in_file(
                    offset,
                    // saturation to u16 for incorrect pdf file, they may use 65536
                    // which out of the range of u16
                    generation.clamp(0, u16::MAX as u32) as u16,
                    ty == b"n",
                )
            },
        ),
    );
    let group = tuple((record_count_parser, many0(record_parser)));
    let parser = fold_many1(group, Vec::new, |mut table, ((start, count), entries)| {
        assert_eq!(count as usize, entries.len());
        for (i, entry) in entries.into_iter().enumerate() {
            table.push((start + u32::try_from(i).unwrap(), entry));
        }
        table
    });

    preceded(context("xref", ws_terminated(tag(b"xref"))), parser)(buf)
}

fn parse_startxref(buf: &[u8]) -> ParseResult<u32> {
    preceded(ws_terminated(tag(b"startxref")), ws_terminated(u32))(buf)
}

// Assumes buf start from xref
fn parse_frame(buf: &[u8]) -> ParseResult<(Dictionary, XRefSection)> {
    map(
        alt((
            tuple((
                context("xref table", parse_xref_table),
                context("trailer", parse_trailer),
            )),
            parse_xref_stream,
        )),
        |(xref_table, trailer)| (trailer, xref_table),
    )(buf)
}

pub fn parse_frame_set(input: &[u8]) -> ParseResult<FrameSet> {
    fn get_prev(frame: &Frame) -> Option<i32> {
        frame.trailer.get(&sname("Prev")).map(|o| o.int().unwrap())
    }

    let mut frames = Vec::new();
    let (buf, _) = context("move to xref", new_r_to_tag(b"startxref"))(input)?;
    let (_, pos) = context("locate frame pos", parse_startxref)(buf)?;
    info!("frame pos: {}", pos);
    let (_, frame) = parse_frame(&input[pos as usize..])?;
    let frame = Frame::new(pos, frame.0, frame.1);
    let mut prev = get_prev(&frame);
    frames.push(frame);

    while let Some(pos) = prev {
        info!("frame pos: {}", pos);
        let buf = &input[pos as usize..];
        let (_, frame) = parse_frame(buf)?;
        let frame = Frame::new(pos as u32, frame.0, frame.1);
        prev = get_prev(&frame);
        frames.push(frame);
    }

    Ok((&input[..0], frames))
}

#[cfg(test)]
mod tests;
