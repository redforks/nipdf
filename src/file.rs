//! Contains types of PDF file structures.
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct Header<'a>(&'a [u8]);

impl<'a> Header<'a> {
    pub fn new(buf: &'a [u8]) -> Self {
        Self(buf)
    }

    pub fn ver(&self) -> &str {
        std::str::from_utf8(self.0).unwrap()
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct Tail {
    xref_pos: u32,
}

impl Tail {
    pub fn new(xref_pos: u32) -> Self {
        Self { xref_pos }
    }

    pub fn xref_pos(&self) -> u32 {
        self.xref_pos
    }
}
