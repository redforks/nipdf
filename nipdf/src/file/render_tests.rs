//! Test page render result using `insta` to ensure that the rendering result is not changed.
//! This file checks file pdfreference1.0.pdf
use crate::file::{File, RenderOptionBuilder};
use anyhow::Result as AnyResult;
use insta::assert_ron_snapshot;
use md5::{Digest, Md5};
use std::path::Path;

fn decode_file_page(path: &str, page_no: usize) -> AnyResult<String> {
    let path = Path::new(env!("CARGO_MANIFEST_DIR")).join(path);
    let buf = std::fs::read(&path)?;
    let f = File::parse(buf, "", "").unwrap_or_else(|e| panic!("failed to parse {path:?}: \n{e}"));
    let resolver = f.resolver()?;
    let catalog = f.catalog(&resolver)?;
    let pages = catalog.pages()?;
    let page = &pages[page_no];
    let option = RenderOptionBuilder::new().zoom(1.5);
    let bytes = page.render(option)?.into_vec();
    let hash = Md5::digest(&bytes[..]);
    Ok(hex::encode(hash))
}

/// Render page to image, and returns its md5 hash converted to hex
fn decode_page(page_no: usize) -> AnyResult<String> {
    decode_file_page("sample_files/normal/pdfreference1.0.pdf", page_no)
}

#[test]
fn clip_mask() {
    assert_ron_snapshot!(&decode_page(141).unwrap());
}

#[test]
fn mask_image() {
    assert_ron_snapshot!(&decode_page(163).unwrap());
}

#[test]
fn pattern_color() {
    assert_ron_snapshot!(
        &decode_file_page("sample_files/normal/SamplePdf1_12mb_6pages.pdf", 5).unwrap()
    );
}

#[test]
fn form() {
    assert_ron_snapshot!(&decode_file_page("sample_files/xobject/form.pdf", 0).unwrap());
}

#[test]
fn form_ctm() {
    assert_ron_snapshot!(
        &decode_file_page("../../pdf/ICEpower125ASX2_Datasheet_2.0.pdf", 4).unwrap()
    );
}

#[test]
fn radial_shade() {
    assert_ron_snapshot!(
        &decode_file_page("sample_files/bizarre/pdfReferenceUpdated.pdf", 809).unwrap()
    )
}

#[test]
fn axial_shade() {
    // TODO: find a sample page contains PaintShading("axial-shade") operation
    assert_ron_snapshot!(&decode_file_page("../../pdf/code.pdf", 619).unwrap())
}

#[test]
fn type0_cid_font() {
    assert_ron_snapshot!(
        &decode_file_page("sample_files/bizarre/pdfReferenceUpdated.pdf", 1013).unwrap()
    )
}

#[test]
fn standard_14_font_alias_name() {
    // Font name: TimesNewRomanPSMT alias of Times-Roman, see normalize_font_name()
    assert_ron_snapshot!(&decode_file_page("../../pdf/code.pdf", 620).unwrap())
}

#[test]
fn image_mask_cal_rgb_index_color_space() {
    // test paint image has mask, and its color space is Indexed to CalRGB,
    // image stream processed with Predicator
    assert_ron_snapshot!(&decode_file_page("sample_files/filters/predictor.pdf", 0).unwrap())
}

#[test]
fn decrypt_aes_revion3() {
    assert_ron_snapshot!(&decode_file_page("../../pdf/5176.CFF.pdf", 0).unwrap());
}

#[test]
fn ttf_font_cmap_trimmed_table_mapping() {
    // font used in graph, its cmap table uses format TrimmedTableMapping
    // that ttf-parser glyph_index() don't work, see `TTFParserFontOp::char_to_gid()`
    assert_ron_snapshot!(
        &decode_file_page("pdf.js/web/compressed.tracemonkey-pldi-09.pdf", 9).unwrap()
    )
}

#[test]
fn axial_shade_with_sample_function() {
    assert_ron_snapshot!(
        &decode_file_page("pdf.js/web/compressed.tracemonkey-pldi-09.pdf", 10).unwrap()
    )
}

#[test]
fn todo_rotate_n_encrypt_alg2() {
    // page rotate
    // encrypt algorithm 2 (Algorithm::Key40AndMore)
    // todo: ForceBold font flag should render glyph bolder
    assert_ron_snapshot!(&decode_file_page("../../pdf/avr-1507-owners-manual-en.pdf", 10).unwrap())
}

#[test]
fn todo_radius_pattern() {
    // todo: tiny_skia not support end radius, the page has both start and end radius rendered
    // incorrectly
    assert_ron_snapshot!(&decode_file_page("sample_files/bizarre/PDF32000_2008.pdf", 745).unwrap())
}

#[test]
fn todo_glyph_encoding_problem() {
    // todo: incorrect bullet glyph rendered, possible because ttf-parser glyph_index() returned
    // wrong glyph index
    assert_ron_snapshot!(&decode_file_page("sample_files/bizarre/PDF32000_2008.pdf", 159).unwrap())
}
