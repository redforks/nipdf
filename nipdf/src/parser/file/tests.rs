use super::*;
use crate::file::open_test_file;
use insta::assert_debug_snapshot;
use test_case::test_case;
use test_log::test;

#[test]
fn parse_file_header() {
    let buf = b"%PDF-1.7\n%comment\n";
    let (buf, header) = parse_header(buf).unwrap();
    assert_eq!(buf, b"");
    assert_eq!(header, "1.7");
}

#[test_case(None, b"hello", b"world"; "not exist")]
#[test_case(Some(0), b"hello", b"hello"; "matches")]
#[test_case(Some(1), b"\nhello", b"hello"; "after newline")]
#[test_case(Some(1), b"\nhello\n", b"hello"; "end with newline")]
#[test_case(Some(2), b"\r\nhello\r\n", b"hello"; "CRLF")]
#[test_case(Some(4), b"foo\nfoo\nbar", b"foo"; "from end")]
#[test_case(None, b"abc-foo", b"foo"; "not the whole line")]
fn test_r_find_start_object_tag(exp: Option<usize>, buf: &[u8], tag: &[u8]) {
    assert_eq!(exp, r_find_start_object_tag(buf, tag));
}

#[test]
fn test_parse_trailer() {
    let buf = b"trailer\n<< /Size 1 >>\nstartxref\n1234\n%%EOF";
    assert_debug_snapshot!(parse_trailer(buf).unwrap());
}

#[test_case(b"xref\n1 0\n", "empty")]
#[test_case(
    b"xref\n1 2\n0000000000 00000 n \n0000000010 00000 n \n",
    "two entries"
)]
fn test_parse_xref_table(buf: impl AsRef<[u8]>, name: &str) {
    assert_debug_snapshot!(name, parse_xref_table(buf.as_ref()).unwrap());
}

#[test]
fn test_parse_frame() {
    assert_debug_snapshot!(
        parse_frame(
            b"xref
1 2
0000000000 00000 n
0000000010 00000 n
trailer
<< /Size 1 >>
startxref
1234
%%EOF
"
        )
        .unwrap()
    );
}

#[test]
fn test_parse_frame_set() {
    assert_debug_snapshot!(
        parse_frame_set(
            b"%PDF-1.7
xref
1 1
0000000000 00000 n
trailer
<< /Size 1 >>
startxref
9
%%EOF
xref
1 1
0000000000 00000 n
trailer
<< /Prev 9 >>
startxref
77
%%EOF
"
        )
        .unwrap()
    );
}

#[test]
fn read_xref_stream() {
    let f = open_test_file("sample_files/file-structure/xref-stream.pdf");
    let resolver = f.resolver().unwrap();

    // assert object in file
    assert_eq!(
        sname("Catalog"),
        resolver
            .resolve(1)
            .unwrap()
            .as_dict()
            .unwrap()
            .get(&sname("Type"))
            .unwrap()
            .name()
            .unwrap()
    );

    // assert object in stream
    assert_eq!(
        1,
        resolver
            .resolve(2)
            .unwrap()
            .as_dict()
            .unwrap()
            .get(&sname("Count"))
            .unwrap()
            .int()
            .unwrap()
    );
}

#[test]
fn read_xref_stream_has_prev() {
    let f = open_test_file("pdf.js/test/pdfs/160F-2019.pdf");
    // get resolver without error
    f.resolver().unwrap();
}
