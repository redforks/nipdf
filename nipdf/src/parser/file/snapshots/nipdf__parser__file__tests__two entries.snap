---
source: nipdf/src/parser/file/tests.rs
expression: parse_xref_table(buf.as_ref()).unwrap()
---
(
    [],
    [
        (
            1,
            InFile(
                FilePos(
                    0,
                    0,
                    true,
                ),
            ),
        ),
        (
            2,
            InFile(
                FilePos(
                    10,
                    0,
                    true,
                ),
            ),
        ),
    ],
)
